Only in a: gcc-12.patch
diff -r -U3 a/src/libc/ansi/stdio/makefile b/src/libc/ansi/stdio/makefile
--- a/src/libc/ansi/stdio/makefile      2001-05-21 19:25:49.000000000 +0200
+++ b/src/libc/ansi/stdio/makefile      2023-01-14 14:31:35.361845387 +0100
@@ -66,3 +66,6 @@
 SRC += vsnprntf.c

 include $(TOP)/../makefile.inc
+
+doscan.o: EXTRA_CFLAGS= -Wno-use-after-free
+rename.o: EXTRA_CFLAGS= -Wno-use-after-free
diff -r -U3 a/src/libc/ansi/stdlib/makefile b/src/libc/ansi/stdlib/makefile
--- a/src/libc/ansi/stdlib/makefile     2015-05-02 06:32:11.000000000 +0200
+++ b/src/libc/ansi/stdlib/makefile     2023-01-14 14:31:35.371845403 +0100
@@ -37,6 +37,6 @@
 # FIXME: This is really only needed for gcc-5 .
 #        Unfortunatelly GCC version is not yet checked
 #        here.
-nmalloc.o: EXTRA_CFLAGS := -fno-builtin-malloc
+nmalloc.o: EXTRA_CFLAGS := -fno-builtin-malloc -Wno-free-nonheap-object

 include $(TOP)/../makefile.inc
diff -r -U3 a/src/libc/dos/process/makefile b/src/libc/dos/process/makefile
--- a/src/libc/dos/process/makefile     2023-01-14 14:52:48.651858329 +0100
+++ b/src/libc/dos/process/makefile     2023-01-14 14:31:35.381845420 +0100
@@ -14,4 +14,4 @@

 include $(TOP)/../makefile.inc

-dosexec.o: EXTRA_CFLAGS += -Wno-stringop-overflow
+dosexec.o: EXTRA_CFLAGS += -Wno-stringop-overflow -Wno-stringop-truncation
diff -r -U3 a/src/libemu/src/makefile b/src/libemu/src/makefile
--- a/src/libemu/src/makefile   2015-05-02 06:32:33.000000000 +0200
+++ b/src/libemu/src/makefile   2023-01-14 14:31:35.401845452 +0100
@@ -16,7 +16,7 @@
 # FIXME: fix emu387.cc instead to avoid need for -fno-strict-aliasing

 %.o : %.cc
-	$(XNOPGGPP) -c -fno-exceptions -fno-strict-aliasing $<
+	$(XNOPGGPP) -c -fno-exceptions -fno-strict-aliasing -Wno-register $<

 %.o : %.c
        $(XNOPGGCC) -c $<
diff -r -U3 a/src/libm/math/makefile b/src/libm/math/makefile
--- a/src/libm/math/makefile    2015-07-18 16:07:52.000000000 +0200
+++ b/src/libm/math/makefile    2023-01-14 14:31:35.411845468 +0100
@@ -224,6 +224,9 @@
 chew.exe: chew.c
        $(GCC) -O2 -o $@ $<

+k_rem_pio2.o: EXTRA_CFLAGS += -Wno-maybe-uninitialized
+kf_rem_pio2.o: EXTRA_CFLAGS += -Wno-maybe-uninitialized
+
 clean ::
        -$(MISC) rm *.def *.d chew.exe targetdep.texi